kotlin/codegen/src/test/java/com/squareup/moshi/kotlin/codegen/MetadataTest.kt

```diff
@@ -0,0 +1,187 @@
+    val fooParam = classData.constructorData!!.parameters[0]
+    assertThat(fooParam.name).isEqualTo("foo")
+    assertThat(fooParam.type).isEqualTo(String::class.asClassName())
+    assertThat(fooParam.isVarArg).isFalse()
+    val barParam = classData.constructorData.parameters[1]
+    assertThat(barParam.name).isEqualTo("bar")
+    assertThat(barParam.type).isEqualTo(IntArray::class.asClassName())
+    assertThat(barParam.isVarArg).isTrue()
+    assertThat(barParam.varargElementType).isEqualTo(INT)
+  }
+
+  class ConstructorClass(val foo: String, vararg bar: Int) {
+    constructor(bar: Int) : this("defaultFoo") // Secondary constructors are ignored
+  }
+
+  @Test
+  fun superType() {
+    val classData = SuperType::class.loadClassData()
+
+    assertThat(classData.superTypes).hasSize(1)
+    assertThat(classData.superTypes[0]).isEqualTo(BaseType::class.asClassName())
+  }
+
+  abstract class BaseType
+  class SuperType : BaseType()
+
+  @Test
+  fun properties() {
+    val classData = Properties::class.loadClassData()
+
+    assertThat(classData.properties).hasSize(4)
+
+    val fooProp = classData.properties.find { it.name == "foo" } ?: throw AssertionError("Missing foo property")
+    assertThat(fooProp.type).isEqualTo(String::class.asClassName())
+    assertThat(fooProp.hasSetter).isFalse()
+    val barProp = classData.properties.find { it.name == "bar" } ?: throw AssertionError("Missing bar property")
+    assertThat(barProp.type).isEqualTo(String::class.asClassName().asNullable())
+    assertThat(barProp.hasSetter).isFalse()
+    val bazProp = classData.properties.find { it.name == "baz" } ?: throw AssertionError("Missing baz property")
+    assertThat(bazProp.type).isEqualTo(Int::class.asClassName())
+    assertThat(bazProp.hasSetter).isTrue()
+    val listProp = classData.properties.find { it.name == "aList" } ?: throw AssertionError("Missing baz property")
+    assertThat(listProp.type).isEqualTo(List::class.parameterizedBy(Int::class))
+    assertThat(listProp.hasSetter).isTrue()
+  }
+
+  class Properties {
+    val foo: String = ""
+    val bar: String? = null
+    var baz: Int = 0
+    var aList: List<Int> = emptyList()
+  }
+
+  @Test
+  fun companionObject() {
+    val classData = CompanionObject::class.loadClassData()
+    assertThat(classData.companionObjectName).isEqualTo("Companion")
+    val namedClassData = NamedCompanionObject::class.loadClassData()
+    assertThat(namedClassData.companionObjectName).isEqualTo("Named")
+  }
+
+  class CompanionObject {
+    companion object
+  }
+
+  class NamedCompanionObject {
+    companion object Named
+  }
+
+  @Test
+  fun generics() {
+    val classData = Generics::class.loadClassData()
+
+    assertThat(classData.typeVariables).hasSize(3)
+    val tType = classData.typeVariables[0]
+    assertThat(tType.name).isEqualTo("T")
+    assertThat(tType.reified).isFalse()
+    assertThat(tType.variance).isNull() // we don't redeclare out variance
```

Comment:
Can you explain this? What does "redeclaring" mean in this context?

Reply:
It's in reference to https://github.com/square/moshi/blob/78821bbc8041bbc1735aa3e1f41629c8d230ff7c/kotlin/codegen/src/main/java/com/squareup/moshi/kotlin/codegen/metadata.kt#L113

Open to wording suggestions :)